[DEM] Migrate to processes and several performance improvements.

maceligueta opened the pull request:
**Description**
Trying to increase performance skipping methods which are useless in some cases.

@joaquinirazabal , it would be great to skip the computation of forces on walls when the wall is totally fixed (all 6 dofs). We could check this solid by solid, and maybe we would save a lot of computation. Unless the user wants to print the pressure on the walls or something similar...

Commit:
some functions can skip their part in some circumstances

philbucher commented on applications/DEMApplication/python_scripts/dem_default_input_parameters.py:
```diff
@@ -40,6 +40,7 @@ def GetDefaultInputParameters():
             "RotationOption"                   : true,
             "CleanIndentationsOption"          : false,
             "RemoveBallsInEmbeddedOption"      : false,
+            "ComputeReactionsOnFixedParticlesOption" : false,
```
```suggestion
            "compute_reactions_on_fixed_particles" : false,
```

more consistent with rest of Kratos

Commit:
style of new parameter more similar to the rest of kratos

joaquinirazabal committed:
Avoid force computation on rigid elements

joaquinirazabal committed:
Merge branch 'master' into dem/performance-improvements

joaquinirazabal committed:
Minor

joaquinirazabal committed:
CalculateConditionsRHSAndAdd only when it is needed for rigid bodies

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_strategy.cpp:
```diff
@@ -913,54 +915,66 @@ namespace Kratos {
-        ProcessInfo& r_process_info = GetFemModelPart().GetProcessInfo();
-        const ProcessInfo& r_const_process_info = GetFemModelPart().GetProcessInfo();
 
+        for (ModelPart::SubModelPartsContainerType::iterator sub_model_part = GetFemModelPart().SubModelPartsBegin(); sub_model_part != GetFemModelPart().SubModelPartsEnd(); ++sub_model_part) {
```
behold the power of `auto` :D

```suggestion
        for (auto& submp : GetFemModelPart().SubModelParts()) {
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_strategy.cpp:
```diff
@@ -913,54 +915,66 @@ namespace Kratos {
-            Vector rhs_cond;
-            Vector rhs_cond_elas;
-        };
+            ModelPart& submp = *sub_model_part;
```
```suggestion
```

philbucher commented on applications/DEMApplication/custom_strategies/strategies/explicit_solver_strategy.cpp:
```diff
@@ -913,54 +915,66 @@ namespace Kratos {
-            DEMWall* p_wall = dynamic_cast<DEMWall*> (&(rCondition));
-            p_wall->CalculateElasticForces(rTLS.rhs_cond_elas, r_process_info);
+            ElementsArrayType& pElements = submp.GetCommunicator().LocalMesh().Elements();
+            ElementsArrayType::iterator it = pElements.ptr_begin();
```
note that this segfaults if there are no elements!
=> I would add at least a debug check

maceligueta replied:
I think this check should be in Release too. I mean, it has a very small cost, because it is not an inner loop, and it can save a lot of time. No need to re-compile in Debug, for example. Or no need for external users to send us an email with "this is not working, guys".

joaquinirazabal committed:
Addressing Philipp Bucher's suggestions

joaquinirazabal committed:
Change error from debug to release

maceligueta commented on applications/DEMApplication/custom_elements/rigid_body_element.cpp:
```diff
@@ -137,6 +137,20 @@ namespace Kratos {
+            if (central_node.Is(DEMFlags::FIXED_VEL_X) && central_node.Is(DEMFlags::FIXED_VEL_Y) && central_node.Is(DEMFlags::FIXED_VEL_Z) && central_node.Is(DEMFlags::FIXED_ANG_VEL_X) && central_node.Is(DEMFlags::FIXED_ANG_VEL_Y) && central_node.Is(DEMFlags::FIXED_ANG_VEL_Z)) {
+                if (rigid_body_element_sub_model_part.Has(COMPUTE_FORCES_ON_THIS_RIGID_ELEMENT)) {
+                    if (rigid_body_element_sub_model_part[COMPUTE_FORCES_ON_THIS_RIGID_ELEMENT]== 1) {
+                        central_node.Set(DEMFlags::COMPUTE_REACTIONS, true);
```
Doing this in Initialize is kind of dangerous. What happens if the user releases the boundary conditions later on? We should probably check this at every time step.

joaquinirazabal replied:
Ok, now it is done in the strategy, at each time step. I also consider that if the submodelpart is FORCE_INTEGRATION_GROUP, forces should be calculated as we need them.

maceligueta replied:
True, we should add this, too.

joaquinirazabal committed:
Move reactions computation condition to the strategy

joaquinirazabal committed:
Fix some tests

Commit:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
updated loop

Commit:
Merge branch 'dem/performance-improvements' of https://github.com/KratosMultiphysics/Kratos into dem/performance-improvements

Commit:
trying to fix tests, first version of the code

farrufat-cimne committed:
fix dem tests

Commit:
SEARCH_CONTROL back to int

farrufat-cimne committed:
some intermediate changes

and updated the kinematic test with analytic results to verify when required.

farrufat-cimne committed:
update analytics test

farrufat-cimne committed:
removed legacy

farrufat-cimne committed:
update json in constituvelaws test

test does not work and it is not included in the list

farrufat-cimne committed:
update

farrufat-cimne committed:
revised dem inlet 2d test

farrufat-cimne committed:
revised restitution test

lowered tolerance. value is exact the closer is to 1.0

farrufat-cimne committed:
migrated test 2D impact to process

revised values and tolerance

farrufat-cimne committed:
revised tests

moved conditions to process

farrufat-cimne committed:
revised more tests

moved to process

farrufat-cimne committed:
removed legacy

removed free body motion flag
removed applyprescribed and initialconditions
removed deprecated defs
removed moveallmeshes

farrufat-cimne committed:
adding rigid body option

farrufat-cimne committed:
fixed 3D continuum test

farrufat-cimne committed:
fixed control module test

farrufat-cimne committed:
revised friction decay test

farrufat-cimne committed:
revised history dep tests

farrufat-cimne committed:
fixed cluster test

farrufat-cimne committed:
revised test schemes

all tests working ok with processes

farrufat-cimne committed:
added check for rigid body variables

revised all tests

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
removed fixed mesh option

and changed function to check smp variables

farrufat-cimne committed:
removed legacy variables in SDEM

farrufat-cimne committed:
fix test in debug

farrufat-cimne committed:
updated swimming dem tests

some tests were failing in dbg

farrufat-cimne committed:
updated more swimming dem tests

farrufat-cimne committed:
added test for omp

results must be exactly the same for different omp setups

farrufat-cimne committed:
removing self on classmethods

it makes no sense

farrufat-cimne committed:
completly removed classmethod deco

Commit:
removing member variable which duplicates location of data

farrufat-cimne committed:
removing identifier variable

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
removed deprecated variable

farrufat-cimne committed:
split force and moments processes

first step

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
forces and moments split

farrufat-cimne committed:
update

farrufat-cimne committed:
update

analytics test

farrufat-cimne committed:
fix initial interval

farrufat-cimne committed:
update tests

all tests ok

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
swimming tests update initial conditions

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
Merge branch 'master' into dem/performance-improvements

farrufat-cimne committed:
Merge branch 'master' into temp

farrufat-cimne committed:
Merge branch 'temp' into dem/performance-improvements

farrufat-cimne committed:
modified process reading to standards

farrufat-cimne committed:
updated sdem tests

farrufat-cimne committed:
interval standards

farrufat-cimne committed:
updated defaults for processes

farrufat-cimne committed:
added test

added small test for fem dofs and interaction with dem when fem is fully or partially constrained
